gtda/plotting/persistence_diagrams.py

```diff
@@ -138,3 +141,126 @@ def plot_diagram(diagram, homology_dimensions=None, plotly_params=None):
+        downwards sweep) and the first two contain (birth, death) pairs to be
+        used as coordinates in the two-dimensional plot.
+
+    homology_dimensions : list of int or None, optional, default: ``None``
+        Homology dimensions which will appear on the plot. If ``None``, all
+        homology dimensions which appear in `diagram` will be plotted.
+
+    plotly_params : dict or None, optional, default: ``None``
+        Custom parameters to configure the plotly figure. Allowed keys are
+        ``"traces"`` and ``"layout"``, and the corresponding values should be
+        dictionaries containing keyword arguments as would be fed to the
+        :meth:`update_traces` and :meth:`update_layout` methods of
+        :class:`plotly.graph_objects.Figure`.
+
+    Returns
+    -------
+    fig : :class:`plotly.graph_objects.Figure` object
+        Figure representing the persistence diagram.
+
+    """
+    # TODO: increase the marker size
+    if homology_dimensions is None:
+        homology_dimensions = np.unique(diagram[:, 2])
+
+    diag = diagram[diagram[:, 0] != diagram[:, 1]]
+    diag_no_dims = diag[:, :2]
+    posinfinite_mask = np.isposinf(diag_no_dims)
+    neginfinite_mask = np.isneginf(diag_no_dims)
+    max_val = np.max(np.where(posinfinite_mask, -np.inf, diag_no_dims))
+    min_val = np.min(np.where(neginfinite_mask, np.inf, diag_no_dims))
+    parameter_range = max_val - min_val
+    extra_space_factor = 0.02
+    extra_space = extra_space_factor * parameter_range
+    min_val_display = min_val - extra_space
+    max_val_display = max_val + extra_space
+
+    subplot_titles = ["Same sweep", "Different sweep"]
+    fig = make_subplots(rows=2, cols=1,
+                        subplot_titles=subplot_titles)
+
+    for row in [1, 2]:
+        fig.add_trace(go.Scatter(
+            x=[min_val_display, max_val_display],
+            y=[min_val_display, max_val_display],
+            mode="lines",
+            line={"dash": "dash", "width": 1, "color": "black"},
+            showlegend=False,
+            hoverinfo="none"
+            ), row=row, col=1)
+
+    for i, sweep in enumerate([1., -1.]):
+        diag_sweep = diag[diag[:, 3] == sweep]
+        for j, dim in enumerate(homology_dimensions):
+            name = f"H{int(dim)}" \
+                if dim != np.inf else "Any homology dimension"
+            subdiag = diag_sweep[diag_sweep[:, 2] == dim]
+            unique, inverse, counts = np.unique(
+                subdiag, axis=0, return_inverse=True, return_counts=True
+                )
+            hovertext = [
+                f"{tuple(unique[unique_row_index][:2])}" +
+                (
+                    f", multiplicity: {counts[unique_row_index]}"
+                    if counts[unique_row_index] > 1 else ""
+                )
+                for unique_row_index in inverse
+                ]
+            fig.add_trace(go.Scatter(
+                x=subdiag[:, 0], y=subdiag[:, 1], mode="markers",
```

Comment:
Shouldn't x,y be `unique`?

Reply:
This is the same as in the non-extended version. It means we duplicate a bunch of hovertexts too.  Maybe it could be done better.

Reply:
Sorry, i misunderstood how the `inverse` in np.unique works. All good :)